api/server/router/image/image_routes.go

```diff
@@ -222,20 +222,23 @@ func (s *imageRouter) getImagesJSON(ctx context.Context, w http.ResponseWriter,
 		return err
 	}
 
-	imageFilters, err := filters.FromJSON(r.Form.Get("filters"))
-	if err != nil {
-		return err
+	opts := types.ImageListOptions{
+		All: httputils.BoolValue(r, "all"),
 	}
-
-	version := httputils.VersionFromContext(ctx)
-	if versions.LessThan(version, "1.41") {
-		filterParam := r.Form.Get("filter")
-		if filterParam != "" {
-			imageFilters.Add("reference", filterParam)
+	if versions.LessThan(httputils.VersionFromContext(ctx), "1.41") {
+		if s := r.Form.Get("filter"); s != "" {
+			opts.Filters = filters.NewArgs(filters.Arg("reference", s))
+		}
+	} else {
+		filters, err := filters.FromJSON(r.Form.Get("filters"))
+		if err != nil {
+			return err
 		}
```

Comment:
This doesn't look correct; in the old code;

1. the "filters" option would be applied
2. if API < 1.41, and the `filter` (singular) option was set, added to the list of filters as `reference` filter

I think `2.` was because before `--filter` was added, it was possible to filter images by name (reference) only as a positional argument (`docker images <search>`); when the `--filter` flag was added, code was added to merge the existing (positional) filter argument with the new `--filter` options. Iterating over the "blame" view history, I see the `filter` option was implemented at least 6 years ago; https://github.com/moby/moby/blame/23750fb80280e6770590b0ea30781c43f42e430d/api/server/image.go#L367, well before API v1.41 (looks like it was originally added in https://github.com/moby/moby/commit/5f3812ec97f9a951bceb482fdcd355b6cbc1b113, as part of https://github.com/moby/moby/pull/4430

At the API level, the old `filter` query parameter was deprecated in favor of `filters` parameter (equivalent to `--filter reference=xxx`), see https://github.com/moby/moby/pull/27872

I'm a bit confused now, as deprecated the `filter` query-arg (for API v1.28 and above), but I see I later opened https://github.com/moby/moby/pull/40491 to deprecate it _again_. When I opened that PR, it looks like the API v1.28 check was removed (??), so I used API v1.41 instead. Will have to find when that v1.28 check was removed (as that would've "un-deprecated" the option?

Reply:
Ah; found where it was removed; https://github.com/moby/moby/pull/31844

Reason at the time was that it was "to be deprecated after two releases", and it was yet unknown what API version that would be, so the check was replaced by a "TODO" instead, which wasn't picked up until v1.41

Reply:
The logic was changed due to your comment here https://github.com/moby/moby/pull/42531#discussion_r659820126, as we discussed offline.
I will drop https://github.com/moby/moby/pull/42531/commits/7821e787813d356e018bd189e59a143b69e0516f, such API changes can be handled in another PR, where such changes are in scope.